### ISSUE TITLE
Research workflow management/orchestration tools

### ISSUE THREAD

**Issue by jtotoole** [mediacloud/backend#760]
In preparation for sitemap work and other upgrades, research options for orchestration tooling

@pypt + @dsjen, plz chime in at your convenience regarding the specific requirements/features I ought to be looking for

**Comment by dsjen** [comment]
Thanks for taking this on and for reaching out! I'm going to stay high level as I don't know too much about the nitty-gritties of the system. Drawing from the ["ilities"](https://en.wikipedia.org/wiki/List_of_system_quality_attributes), I think the new system should at least have high marks in:

* fault-tolerance / resilience -- one thing going wrong doesn't necessarily bring down the whole system
* debug-ability -- there enough transparency so that we can see what's going wrong with the system and also what's going wrong with the job
* responsiveness -- it's relatively performant
* compatibility -- works for managing our processes (e.g. we don't have to rewrite all the jobs)
* sustainability -- it's not going to go away anytime soon, has a support community, and is maintained by more than a sole developer
* credibility -- it's used by others

Nice to haves:
* learn-ability -- it makes sense and it's a huge pain to learn how to use
* extensibility -- well suited for future stuff we might do (whatever that is)
* flexibility -- (this is the best "ility" to represent what I was thinking) it can be slotted into our system iteratively. This is low on the list of "nice to haves" for me, but would be useful for slowly shifting jobs over to the new system.

**Comment by pypt** [comment]
A lot of what we do on the backend happens in some sort of a pipeline. We called it an ETL pipeline in our original job ad, so maybe it's even called that (I wouldn't know, they kicked me out from CS101 :)), but in a general sense one of the applications that we run comes up with some inputs (e.g. finds new stories on the web) and passes it so some other app for it to process in parallel (e.g. some sort of a fetcher process). That other application does its thing with its input (so many instances of the same application run in parallel to go through all the inputs in one go), in turn come up with some more stuff for yet other apps to do, those apps do their thing, and so on and so on.

How this tree-like-but-not-quite-a-tree thing is currently implemented is through simple queues: once `app1` comes up with some stuff for `app2` to do, it adds it to `app2`'s queue, then whenever `app2` has a minute to get busy with something new, it polls said queue (or, rather, RabbitMQ lets the app know that there's a new message waiting, but that's beyond the point).

This approach works for simple processes that we run but becomes increasingly painful for more complex / busy ones due to the reasons including and not limited to:

* If job 12352354353 of the app `app1` failed, how do we find out about it?
* If job 12352354353 of the app `app1` failed, how do we find out what exactly has happened without having to grep through log entries of 12352354352 jobs that have run before the failed job?
* What if a job has to run with a delay? What if said delay is in weeks / months?
* What if a job has to wait for other jobs to complete (think map-reduce)?
* What if we run 10k jobs in parallel, and most of them go through fine but 1% pass and we're just fine with 1% error rate?
* What if a single step in the processing chain had a bug, and it has been fixed since, but the whole chain has been running for a week and now we have to wait for another week for the chain to get fully reprocessed to see if our fix worked?
* Where's my shiny Angular-React-Vue-Bootstrap web UI with bright colors which would tell me what's currently happening to my business-critical processing chain?
* ...

We have workarounds for most of this stuff, but they're only that - workarounds - and that makes it really soul crushing to support a system that's made up from a bunch of nooks and crannies.

So yeah, we've tried KISS approach to processing chain with simple queues, but I'm rather convinced now that simple queues don't quite cut it anymore for us, so I'm more than ready to try something new and more advanced.

My dream workflow management system would be able to:

* Run workers in parallel;
    * Use case: a single server can't do everything that we want to do, so we want to employ all 10-ish servers that we have to do our thing;
* Delay job by hours / weeks / months / to a specific date;
    * Use case: we fetch a story, and we wanna find out how much the URL of said story has been shared / commented on Facebook, but if the story is fresh out of the oven (i.e. has been published just minutes ago), we want to wait for a week or so and only then go hit the share / comment count;
* Run **at least** tens of thousands of tasks of the same kind in parallel;
    * Use case: crawler comes up with 10k URLs to fetch and passes then down the chain for some other process to do the actual fetching;
    * This particular requirement is where many of these workflow systems are lacking unfortunately as many of them support `processA` coming up with only tens or hundreds of inputs for `processB` to do;
* Support millions of pending jobs in the queue;
    * Use case: we want to reextract every story we have with the new version of the extractor that we've deployed;
    * Some other workflow systems are not great at keeping gazillions of jobs in their queue as well, and they're not particularly open about that fact in the docs, so one has to implement a sample workflow chain using the tool and try it out;
* Support parameter idempotence (hope I won't have to try pronouncing this word in any of our meetings), i.e. if a job with parameter `X=42` is already in the processing chain, a new job with the same parameter shouldn't be added;
    * Use case: when we're running a topic, we want to have only a single "run" of it at the time;
* "Reduce" a bunch of "mapped" jobs:
    * Use case: `processA` comes up with 10k URLs to fetch for `processB`; `processB` does the actual fetching, and once all (most?) URLs get attempted at being fetched, `processC` gets an aggregate of fetch results for it to decide what to do next with it;
* Pause / resume;
    * Use case: we run some sort of a heavy workload, and we want to pause it for a bit to be able to run a different workload for a bit; later we resume running the original workflow from the point where it got paused at;
* Retries with preserved state (can't some up with a better term);
    * Use case: we run a complex workload made up from a bunch of finicky steps (workers / apps); worker number 42 turns out to have a bug in it and thus the whole chain fails; we fix the bug in worker 42, deploy the fix, and resume the processing chain from the exact same point where it has failed;
    * Use case 2: topic has been running happily up until a point when it wasn't running so happily anymore due to the fact that it hit the 100k stories limit; user emailed us about it, complained, and so we upped their limit and resumed running the topic from the point when it originally failed due to the limit;
* Loop / recursion support
    * Use case: once we submit a podcast episode for transcription, we have to wait for a bit for the transcription to complete; how much exactly to wait is a bit of a guess, so whenever the process polls the API for the transcript, it's possible that the result is not ready yet; so the process should be able to queue an instance of self to run at a later date. One can implement it with retries too but we want to be able for the chain to "go back" by a few steps whenever it deems it necessary to do so
    * This (probably) doesn't play well with DAGs (e.g. Airflow or Prefect), but I'm not sure
* Logging of individual steps (and - if possible - the whole chain for a specific input)
    * Use case: whenever one of the steps of the topic mapper fails, we want to have a sane way (a few clicks on a webapp, or a command, or something like that) to see the log of the subsubprocess Z of the subprocess Y of process X; and it would be useful to have the ability to see the log of the whole processing chain with a specific input (e.g. topic ID)

Some tools that almost do what we want:

* [Conductor](https://netflix.github.io/conductor/)
* [Cadence](https://cadenceworkflow.io/)
* [Temporal](https://temporal.io/)
* [Zeebe](https://zeebe.io/)
* [Camunda](https://camunda.com/)
* [Prefect](https://www.prefect.io/)

Most of them lack good parallelism / map-reduce support, e.g. step 1 can't come up with 10k tasks for step 2, and / or they can't be reduced back into a single task for step 3. I have notes on some of these tools but I'm already tired of typing :)

Other tools to look at:

* [Airflow](https://airflow.apache.org/)
* [Activiti](https://www.activiti.org/)
* [Toil](https://toil.ucsc-cgl.org/)
* [Arvados](https://arvados.org/)
* [Dagster](https://dagster.io/)
* [Galaxy](https://galaxyproject.org/)
* [Argo](https://github.com/argoproj/argo) - I'm particularly hopeful about this one as it promises to be able to do a lot of stuff that we want, but I've initially dismissed it because it depends on Kubernetes, but now I think that it's almost time for us to outgrow our Docker Swarm shoes and move so K8s

**Comment by pypt** [comment]
^ added a new item to the wishlist on supporting manual user reviews for input data
